workload.go

```diff
@@ -62,32 +76,59 @@ func (w WorkloadConfig) Config() WorkloadConfig {
 }
 
 func NewWorkloadConfigWithDefaults(c WorkloadConfig) WorkloadConfig {
-	if c.RateControl.OuterLoopRate == 0 {
-		c.RateControl.OuterLoopRate = 50
+	if c.Name == "" {
+		panic("must specify workload Name")
 	}
 
-	if c.Visualization.LatencyHistMin == 0 {
-		c.Visualization.LatencyHistMin = 0
+	if c.WorkloadScale <= 0 || c.WorkloadScale >= 1 {
+		panic("WorkloadScale (if specified) must be greater than 0.0 and less than 1.0")
 	}
 
-	if c.Visualization.LatencyHistMax == 0 {
-		c.Visualization.LatencyHistMax = 50000
+	if c.RateControlConfig.EventRate == 0 {
+		c.RateControlConfig.EventRate = 1000
 	}
 
-	if c.Visualization.LatencyHistSize == 0 {
-		c.Visualization.LatencyHistSize = 1000
+	if c.RateControlConfig.Concurrency == 0 {
+		if c.RateControlConfig.MaxEventRatePerWorker == 0 {
+			c.RateControlConfig.MaxEventRatePerWorker = 100
+		}
+		c.RateControlConfig.Concurrency = int(math.Ceil(c.RateControlConfig.EventRate / float64(c.RateControlConfig.MaxEventRatePerWorker)))
+	} else if c.RateControlConfig.Connections >= c.RateControlConfig.Concurrency {
+		c.DatabaseConfig.ConnectionMultiplier = c.RateControlConfig.Connections / c.RateControlConfig.Concurrency
+	} else {
+		c.DatabaseConfig.ConnectionMultiplier = 1
 	}
 
-	if c.RateControl.MaxEventRatePerGoroutine == 0 {
-		c.RateControl.MaxEventRatePerGoroutine = 100 // 2 Event per iteration by default.
+	if c.RateControlConfig.MaxEventRatePerWorker == 0 {
+		c.RateControlConfig.MaxEventRatePerWorker = math.Ceil(c.RateControlConfig.EventRate / float64(c.RateControlConfig.Concurrency))
```

Comment:
Wait do we need this code? You already have it above.

Reply:
That one is for setting a default that in turn gets used to calculate Concurrency (if Concurrency was not specified).
This one is for calculating MaxEventRatePerWorker _from_ Concurrency (if Concurrency _was_ specified).

I'll rearrange this to make the logic more explicit.